### ISSUE TITLE
Kernel/Net Race Condition?

### ISSUE THREAD

**Issue by tryfinally** [SerenityOS/serenity#3226]
https://github.com/SerenityOS/serenity/blob/413db2d6d5a49400d9587c9aa37bedd83968a6dd/Kernel/Net/IPv4Socket.cpp#L50-L53

There is a potential race condition on `s_table` .
Shouldn't checking if null, assignment to and dereferencing s_table be  **_atomic_** ?
I have prepared a _fix_, but am holding the PR.

**Comment by tomuta** [comment]
Not knowing how your fix looks like, I think we should create a reusable singleton class, one that takes a create function and internally will call it one (and once only) to create the singleton (while holding a spinning lock of some sort), and from that point on simply return the created object.

**Comment by 8infy** [comment]
Isn't that what GCC does by default for non-trivially-constructable static objects inside a function? Maybe consider implementing `__cxa_guard_acquire` and `__cxa_guard_release` instead and convert this into a HashTable instance.

**Comment by tomuta** [comment]
> Isn't that what GCC does by default for non-trivially-constructable static objects inside a function? Maybe consider implementing `__cxa_guard_acquire` and `__cxa_guard_release` instead and convert this into a HashTable instance.

I believe that would also require exception handling, which we don't support in the Kernel.

**Comment by tryfinally** [comment]
I did create a reusable singleton

**Comment by tomuta** [comment]
> I did create a reusable singleton

I guess we duplicated efforts here ;-)

**Comment by tryfinally** [comment]
@tomuta   Yup, I should have stated that I testing a solution. But somehow that did not seem fit into an issue report.
Maybe we evolve this into a Singleton and a separate Initialize Once class.

**Comment by ADKaster** [comment]
> > Isn't that what GCC does by default for non-trivially-constructable static objects inside a function? Maybe consider implementing `__cxa_guard_acquire` and `__cxa_guard_release` instead and convert this into a HashTable instance.
> 
> I believe that would also require exception handling, which we don't support in the Kernel.

I did that in my "get rid of libstdc++ entirely" MR that had to get reverted because it broke more issues than it solved. __cxa_guard_acquire and __cxa_guard_release do not require exception handling, they are only related to static variable initialization. Those symbols are already requested by Serenity code (which is why I had to add them), and are provided by libgcc.a
